ciri/consensus/snapshot/local_snapshots/pruning_service.c

```diff
@@ -0,0 +1,333 @@
+                                                          bool *should_branch, bool *should_stop) {
+  retcode_t ret = RC_OK;
+  *should_stop = false;
+  *should_branch = false;
+  collect_transactions_for_pruning_do_func_params_t *params = data;
+
+  if (pack->num_loaded == 0) {
+    return RC_OK;
+  }
+
+  uint64_t current_snapshot_index = transaction_snapshot_index((iota_transaction_t *)pack->models[0]);
+
+  if (current_snapshot_index >= params->current_snapshot_index) {
+    *should_branch = true;
+    if ((ret = hash243_set_add(params->transactions_to_prune, hash)) != RC_OK) {
+      return ret;
+    }
+
+    ERR_BIND_RETURN(
+        tangle_traversal_dfs_to_future(params->tangle, collect_unconfirmed_future_transactions_for_pruning_do_func,
+                                       hash, NULL, &params),
+        ret);
+  }
+
+  if (ret) {
+    *should_stop = true;
+    return ret;
+  }
+
+  return RC_OK;
+}
+
+static void *pruning_service_routine(void *arg) {
+  pruning_service_t *ps = (pruning_service_t *)arg;
+  tangle_t tangle;
+  uint64_t start_timestamp, end_timestamp;
+  uint64_t start_index;
+
+  {
+    connection_config_t db_conf = {.db_path = ps->conf->tangle_db_path};
+
+    if (iota_tangle_init(&tangle, &db_conf) != RC_OK) {
+      log_critical(logger_id, "Failed in initializing db\n");
+      return NULL;
+    }
+  }
+
+  spent_addresses_provider_t sap;
+  {
+    connection_config_t db_conf = {.db_path = ps->spent_addresses_service->conf->spent_addresses_db_path};
+
+    if (ps->spent_addresses_service->conf->spent_addresses_files == NULL) {
+      return RC_OK;
+    }
+
+    if (iota_spent_addresses_provider_init(&sap, &db_conf) != RC_OK) {
+      log_error(logger_id, "Initializing spent addresses database connection failed\n");
+      goto cleanup;
+    }
+  }
+
+  DECLARE_PACK_SINGLE_MILESTONE(milestone, milestone_ptr, milestone_pack);
+  iota_tangle_milestone_load_first(&tangle, &milestone_pack);
+
+  lock_handle_lock(&ps->lock_handle);
+  if (milestone_pack.num_loaded > 0) {
+    ps->last_pruned_snapshot_index = MIN(MIN(0UL, milestone.index - 1), ps->last_pruned_snapshot_index);
+  }
+
+  while (ps->running) {
+    start_index = ps->last_pruned_snapshot_index;
+    start_timestamp = current_timestamp_ms();
+    while (ps->last_pruned_snapshot_index < ps->last_snapshot_index_to_prune) {
+      if (prune_transactions(ps, &tangle, &sap) != RC_OK) {
+        goto cleanup;
+      }
+      cond_handle_timedwait(&ps->cond_pruning_service, &ps->lock_handle, PRUNING_RESCAN_INTERVAL_SHORT_MS);
+    }
+    lock_handle_lock(&ps->lock_handle);
+    end_timestamp = current_timestamp_ms();
+    log_info(logger_id, "Pruning from % " PRIu64 " to % " PRIu64 " took % " PRIu64 " milliseconds\n", start_index,
+             ps->last_pruned_snapshot_index, end_timestamp - start_timestamp);
+    cond_handle_timedwait(&ps->cond_pruning_service, &ps->lock_handle, PRUNING_RESCAN_INTERVAL_MS);
```

Comment:
Isn't 1s a bit too much ?

Reply:
No, it's the `wait` for when pruning is done, next one should start when LS is done, that could take longer actaully